veda-consulting/uk.co.vedaconsulting.mosaico
# Provide better support for high resolution images in mosaico

Overview
----------------------------------------
The PR attempts to provide better support for high resolution images in mosaico https://github.com/veda-consulting/uk.co.vedaconsulting.mosaico/issues/50.

Technical Details
----------------------------------------
A new config has been added for user to specify resize scaling factor:

![image](https://user-images.githubusercontent.com/3448551/79331232-c6393c00-7f12-11ea-8255-ebb9476233b5.png)

When uploading images, the mosaico editor trims it down to very required size (in pixels). 
The solution uses scale factor setting to keep some buffer (2x or 3x) so upscale doesn't look distorted or low resolution.

The new method adjustResizeDimensions() part of graphics interface class, makes it work by modifying the requested resize numbers by editor, before they are used for processing by resize and cover image operations/methods.  

Tests
----------------------------------------
A new e2e test case tests adjustResizeDimensions method for various inputs.

Manual test result:

![image](https://user-images.githubusercontent.com/3448551/79334304-f59e7780-7f17-11ea-89e6-29a6576831ad.png)

If anyone else could test the PR, would be great. cc: @Stoob 

PR Testing Notes
----------------------------------------
- After applying the PR changes, make sure to clear civicrm cache for new configs to appear.
- New configs when set, will get applied to only new uploads. For existing images try to delete and re-upload.
